homeassistant/components/home_connect/coordinator.py

```diff
@@ -0,0 +1,226 @@
+from aiohomeconnect.model import (
+    Event,
+    EventKey,
+    EventMessage,
+    EventType,
+    GetSetting,
+    HomeAppliance,
+    SettingKey,
+    Status,
+    StatusKey,
+)
+from aiohomeconnect.model.error import (
+    EventStreamInterruptedError,
+    HomeConnectApiError,
+    HomeConnectError,
+    HomeConnectRequestError,
+)
+from propcache.api import cached_property
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+from .utils import get_dict_from_home_connect_error
+
+_LOGGER = logging.getLogger(__name__)
+
+type HomeConnectConfigEntry = ConfigEntry[HomeConnectCoordinator]
+
+EVENT_STREAM_RECONNECT_DELAY = 30
+
+
+@dataclass(frozen=True, kw_only=True)
+class HomeConnectApplianceData:
+    """Class to hold Home Connect appliance data."""
+
+    info: HomeAppliance
+    settings: dict[SettingKey, GetSetting]
+    status: dict[StatusKey, Status]
+    events: dict[EventKey, Event] = field(default_factory=dict)
+
+
+class HomeConnectCoordinator(
+    DataUpdateCoordinator[dict[str, HomeConnectApplianceData]]
+):
+    """Class to manage fetching Home Connect data."""
+
+    config_entry: HomeConnectConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: HomeConnectConfigEntry,
+        client: HomeConnectClient,
+    ) -> None:
+        """Initialize."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            config_entry=config_entry,
+            name=config_entry.entry_id,
+        )
+        self.client: HomeConnectClient = client
+        self._home_appliances_event_listeners: dict[
+            str, dict[EventKey, list[Callable[[Event], Coroutine[Any, Any, None]]]]
+        ] = {}
+
+    @cached_property
+    def context_listeners(self) -> dict[tuple[str, EventKey], list[CALLBACK_TYPE]]:
+        """Return a dict of all listeners registered for a given context."""
+        listeners: dict[tuple[str, EventKey], list[CALLBACK_TYPE]] = {}
+        for listener, context in list(self._listeners.values()):
+            assert isinstance(context, tuple)
+            listeners.setdefault(context, []).append(listener)
+        return listeners
+
+    @callback
+    def async_add_listener(
+        self, update_callback: CALLBACK_TYPE, context: Any = None
+    ) -> Callable[[], None]:
+        """Listen for data updates."""
+        remove_listener = super().async_add_listener(update_callback, context)
+        self.__dict__.pop("context_listeners", None)
+
+        def remove_listener_and_invalidate_context_listeners() -> None:
+            remove_listener()
+            self.__dict__.pop("context_listeners", None)
+
+        return remove_listener_and_invalidate_context_listeners
+
+    async def start_event_listener(self) -> None:
```

Comment:
This can be an async callback since we don't await inside. Decorate it with `@callback`.